src/source/split_at.rs

```diff
@@ -0,0 +1,123 @@
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::Duration;
+
+use crate::ChannelCount;
+use crate::SampleRate;
+
+use super::Source;
+use super::TrackPosition;
+
+pub struct SplitAt<S> {
+    shared_source: Arc<Mutex<Option<TrackPosition<S>>>>,
+    active: Option<TrackPosition<S>>,
+    split_duration: Option<Duration>,
+    split_point: Duration,
+    first_span_sample_rate: SampleRate,
+    first_span_channel_count: ChannelCount,
+}
+
+impl<S> SplitAt<S>
+where
+    S: Source,
+    <S as Iterator>::Item: crate::Sample,
+{
+    /// returns two sources, the second is inactive and will return
+    /// none until the first has passed the split_point.
+    pub(crate) fn new(input: S, split_point: Duration) -> [Self; 2] {
+        let shared_source = Arc::new(Mutex::new(None));
+        let first_span_sample_rate = input.sample_rate();
+        let first_span_channel_count = input.channels();
+        let total_duration = input.total_duration();
+        [
+            Self {
+                shared_source: shared_source.clone(),
+                active: Some(input.track_position()),
+                split_duration: Some(split_point),
+                split_point,
+                first_span_sample_rate,
+                first_span_channel_count,
+            },
+            Self {
+                shared_source,
+                active: None,
+                split_duration: total_duration.map(|d| d.saturating_sub(split_point)),
+                split_point: Duration::MAX,
+                first_span_sample_rate,
+                first_span_channel_count,
+            },
+        ]
+    }
+}
+
+impl<S> Iterator for SplitAt<S>
+where
+    S: Source,
+    S::Item: crate::Sample,
+{
+    type Item = <S as Iterator>::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let input = if let Some(active) = self.active.as_mut() {
+            active
+        } else {
+            // did they other stop?
+            let shared = self
+                .shared_source
+                .lock()
+                .expect("audio thread should not panic")
+                .take();
+            self.active = shared;
+            self.active.as_mut()?
+        };
+
+        // There is some optimization potential here we are not using currently.
+        // Calling get_pos once per span should be enough
+        if input.get_pos() < self.split_point {
+            input.next()
+        } else {
+            let source = self.active.take();
+            *self
+                .shared_source
+                .lock()
+                .expect("audio thread should not panic") = source;
```

Comment:
Similarly this could be "lock shared source at the end of first fragment". It could be shorter, but it is better to distinguish this from the expect message in `next`.

Reply:
Its a blocking lock so it can not fail because its already locked, since then it just blocks..

I find: `The audio thread cant panic while first fragment is storing source in lock` clearer, though pretty long.

Reply:
Yes, it is. Maybe you'd have a better idea. I just want the message to be more informative.